FlowCrypt/Functionality/Services/Local Pub Key Services/LocalContactsProvider.swift

```diff
@@ -8,66 +8,77 @@
 
 import Foundation
 import RealmSwift
+import FlowCryptCommon
 
 protocol LocalContactsProviderType: PublicKeyProvider {
-    func updateLastUsedDate(for email: String)
     func searchRecipient(with email: String) async throws -> RecipientWithSortedPubKeys?
     func searchEmails(query: String) -> [String]
-    func save(recipient: RecipientWithSortedPubKeys)
-    func remove(recipient: RecipientWithSortedPubKeys)
-    func updateKeys(for recipient: RecipientWithSortedPubKeys)
+    func save(recipient: RecipientWithSortedPubKeys) throws
+    func remove(recipient: RecipientWithSortedPubKeys) throws
+    func updateKeys(for recipient: RecipientWithSortedPubKeys) throws
     func getAllRecipients() async throws -> [RecipientWithSortedPubKeys]
 }
 
-struct LocalContactsProvider {
-    private let localContactsCache: EncryptedCacheService<RecipientRealmObject>
-    let core: Core
+final class LocalContactsProvider {
+    private let encryptedStorage: EncryptedStorageType
+    private let core: Core
+
+    private lazy var logger = Logger.nested(Self.self)
+
+    private var storage: Realm {
+        encryptedStorage.storage
+    }
 
     init(
         encryptedStorage: EncryptedStorageType,
         core: Core = .shared
     ) {
-        self.localContactsCache = EncryptedCacheService<RecipientRealmObject>(encryptedStorage: encryptedStorage)
+        self.encryptedStorage = encryptedStorage
         self.core = core
     }
 }
 
 extension LocalContactsProvider: LocalContactsProviderType {
-    func updateLastUsedDate(for email: String) {
-        let recipient = find(with: email)
+    func retrievePubKeys(for email: String) -> [String] {
+        guard let object = find(with: email) else { return [] }
 
-        try? localContactsCache.realm.write {
-            recipient?.lastUsed = Date()
+        do {
+            try storage.write {
+                object.lastUsed = Date()
+            }
```

Comment:
For another issue - need to review usage. Just because a public key is retrieved doesn't mean that we should update lastUsed. lastUsed should only be updated when user explicitly uses the recipient, like writing a message to them.